File: OOPs/Constructors/DeepCopyConstructor.java
import java.util.ArrayList;

public class DeepCopyConstructor implements Cloneable {
    String brand;
    ArrayList<String> features;
    
    public DeepCopyConstructor(String brand, ArrayList<String> features) {
        this.brand = brand;
        this.features = new ArrayList<String>(features);
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        Car clonedCar = (Car) super.clone();
        clonedCar.features = new ArrayList<>(this.features);  // Deep copy
        return clonedCar;
    }

    public void displayInfo() {
        System.out.println("Car: " + brand + ", Features: " + features);
    }

    public static void main(String[] args) throws CloneNotSupportedException {
        ArrayList<String> features = new ArrayList<>();
        features.add("Airbags");
        features.add("ABS");

        DeepCopyConstructor car1 = new DeepCopyConstructor("Toyota", features);
        DeepCopyConstructor car2 = (DeepCopyConstructor) car1.clone();  // Deep copy

        // Modify the original object's features
        car1.features.add("Sunroof");

        car1.displayInfo();  // Output: Car: Toyota, Features: [Airbags, ABS, Sunroof]
        car2.displayInfo();  // Output: Car: Toyota, Features: [Airbags, ABS]
    }
}

File: OOPs/Constructors/deep_copy_constructor.cpp
#include <iostream>
#include <vector>
using namespace std;

class Car {
public:
    string brand;
    vector<string>* features;

    // Constructor
    Car(string b, vector<string> f) {
        brand = b;
        features = new vector<string>(f);
    }

    // Deep Copy Constructor
    Car(const Car& other) {
        brand = other.brand;
        features = new vector<string>(*other.features);
    }

    ~Car() {
        delete features;
    }

    void displayInfo() {
        cout << "Car: " << brand << ", Features: ";
        for (const string& feature : *features) {
            cout << feature << " ";
        }
        cout << endl;
    }
};

int main() {
    vector<string> features = {"Airbags", "ABS"};
    Car car1("Toyota", features);

    Car car2 = car1;  // Deep copy

    // Modify the original object's features
    car1.features->push_back("Sunroof");

    car1.displayInfo();  // Output: Car: Toyota, Features: Airbags ABS Sunroof
    car2.displayInfo();  // Output: Car: Toyota, Features: Airbags ABS

    return 0;
}
